### ISSUE TITLE
lint idea: `unnecessary_generic`

### ISSUE THREAD

**Issue by y86-dev** [rust-lang/rust-clippy#14848]
### What it does

Given a type with a generic argument that has a default:
```rust
type Result<T = ()> = core::result::Result<T, MyError>;
```
Usage of this type when specifying the default should not use generics:
```rust
fn foo() -> Result<()> { Ok(()) }
//                ^^^^ unnecessary generic, `()` already is the default
//                hint: use `Result` instead.
```

### Advantage

- Removes duplication of the default value
- Reduces visual clutter
- Reminds people that the type has a default value

### Drawbacks

- when a library adds a default value, one gets this warning when updating
- macros might trigger this involuntarily

### Example

```rust
type Result<T = ()> = core::result::Result<T, MyError>;

fn foo() -> Result<()> {
    Ok(())
}
```

Could be written as:

```rust
type Result<T = ()> = core::result::Result<T, MyError>;

fn foo() -> Result {
    Ok(())
}
```

**Comment by y86-dev** [comment]
I'm not 100% satisfied with the name of the lint "`unnecessary_generic`", but I also couldn't come up with something better.

**Comment by ojeda** [comment]
+1, we had this as https://github.com/Rust-for-Linux/linux/issues/1128 as a kernel-specific `checkpath.pl` "lint", but if Clippy can do this, that would way better: it would cover more projects and more generic types.

**Comment by samueltardieu** [comment]
It has to work with allow/deny sentences, such as in "allow useless default parameters", so `useless_default_parameters` or something like that might be an option.

**Comment by y86-dev** [comment]
> +1, we had this as [Rust-for-Linux/linux#1128](https://github.com/Rust-for-Linux/linux/issues/1128) as a kernel-specific `checkpath.pl` "lint", but if Clippy can do this, that would way better: it would cover more projects and more generic types.

Ah I see, I didn't check, since I think putting this in clippy is the right way, there is nothing kernel-specific or obscure about this lint :)

**Comment by y86-dev** [comment]
> It has to work with allow/deny sentences, such as in "allow useless default parameters", so `useless_default_parameters` or something like that might be an option.

that sounds much better :)

**Comment by Jarcho** [comment]
`explicit_default_arguments` might be better. The implementation has to avoid linting anywhere type inference happens so `check_ty` can't be used.

**Comment by vogelbirb** [comment]
I guess with the feature freeze, it's too late to finish this. I'll still submit the pull request once I can get the time to finish it up.

EDIT: I got a working prototype, but I think I'll hold off on the pull request until the feature freeze ends.

**Comment by sanskarRaghuvanshi** [comment]
Hi @BennoLossin! Is this issue still available? I'd like to work on it.

**Comment by BennoLossin** [comment]
There is an implementation being reviewed in https://github.com/rust-lang/rust-clippy/pull/16123
